src/jobs.rs

```diff
@@ -486,6 +505,156 @@ impl CommitJobInstance {
+            match path {
+                [file] => mtree.lookup(file).map_err(Into::into),
+                [subdir, rest @ ..] => mtree_lookup(
+                    &mtree
+                        .lookup(subdir)?
+                        .1
+                        .ok_or_else(|| "subdirectory not found".to_string())?,
+                    rest,
+                ),
+                [] => Err("no path given".into()),
+            }
+        }
+
+        let mtree_lookup_file = |mtree, path| -> Result<_, Box<dyn Error>> {
+            mtree_lookup(mtree, path)?
+                .0
+                .ok_or_else(|| "file not found".into())
+        };
+
+        let app_id = build_ref.ref_name.split('/').nth(1).unwrap();
+
+        if build_ref.ref_name.starts_with("app/") {
+            if let Some(endpoint) = &config.storefront_info_endpoint {
+                let convert_err =
+                    |e| format!("Failed to fetch storefront info from {}: {}", &endpoint, e);
+
+                /* Fetch the storefront info */
+                let storefront_info = reqwest::blocking::Client::new()
+                    .get(endpoint)
+                    .query(&[("app_id", app_id)])
+                    .send()
+                    .map_err(convert_err)?
+                    .error_for_status()?
+                    .json::<StorefrontInfo>()
+                    .map_err(convert_err)?;
+
+                /* Read the appstream file from the commit */
+                let repo = libostree::Repo::new_for_path(upload_path);
+                repo.open::<Cancellable>(None)?;
+
+                let mtree = libostree::MutableTree::from_commit(&repo, &build_ref.commit)?;
+
+                let appstream_file = mtree_lookup_file(
+                    &mtree,
+                    &[
+                        "files",
+                        "share",
+                        "app-info",
+                        "xmls",
+                        &format!("{}.xml.gz", app_id),
+                    ],
+                ).map_err(|e| format!("Failed to apply storefront info: Could not find the appstream file in the uploaded commit: {}",e))?;
+
+                let (appstream_file, fileinfo, _) =
+                    repo.load_file::<Cancellable>(&appstream_file, None)?;
+
+                let fileinfo = fileinfo.unwrap();
+
+                let appstream_content = appstream_file
+                    .unwrap()
+                    .read_bytes::<Cancellable>(fileinfo.size().try_into().unwrap(), None)?;
+
+                let mut s = String::new();
+                GzDecoder::new(&*appstream_content).read_to_string(&mut s)?;
+
+                storefront_info.validate_storefront_info(app_id, &s)?;
+                let new_appstream = storefront_info.apply_storefront_info(&s)?;
+
+                if new_appstream == s {
+                    /* If the appstream file didn't change, we shouldn't bother rewriting the commit  */
```

Comment:
Makes sense; though a different approach here is to write the commit, but then compare their content checksums and avoid creating a new commit if they're the same.